documentation/proposals/Proposal - Silk.NET 3.X Generic Maths.md

```diff
@@ -0,0 +1,884 @@
+    #endregion
+
+    #region Casts
+    public Vector2D<TOther> As<TOther>() where TOther : INumberBase<TOther>;
+    private Vector2D<TOther> AsChecked<TOther>() where TOther : INumberBase<TOther>;
+    public static explicit operator Vector2D<byte>(Vector2D<T> self);
+    public static explicit operator Vector2D<sbyte>(Vector2D<T> self);
+    public static explicit operator Vector2D<short>(Vector2D<T> self);
+    public static explicit operator Vector2D<ushort>(Vector2D<T> self);
+    public static explicit operator Vector2D<int>(Vector2D<T> self);
+    public static explicit operator Vector2D<uint>(Vector2D<T> self);
+    public static explicit operator Vector2D<long>(Vector2D<T> self);
+    public static explicit operator Vector2D<ulong>(Vector2D<T> self);
+    public static explicit operator Vector2D<Int128>(Vector2D<T> self);
+    public static explicit operator Vector2D<UInt128>(Vector2D<T> self);
+    public static explicit operator Vector2D<Half>(Vector2D<T> self);
+    public static explicit operator Vector2D<float>(Vector2D<T> self);
+    public static explicit operator Vector2D<double>(Vector2D<T> self);
+    public static explicit operator Vector2D<decimal>(Vector2D<T> self);
+    public static explicit operator Vector2D<Complex>(Vector2D<T> self);
+    public static explicit operator Vector2D<BigInteger>(Vector2D<T> self);
+    public static explicit operator checked Vector2D<byte>(Vector2D<T> self);
+    public static explicit operator checked Vector2D<sbyte>(Vector2D<T> self);
+    public static explicit operator checked Vector2D<short>(Vector2D<T> self);
+    public static explicit operator checked Vector2D<ushort>(Vector2D<T> self);
+    public static explicit operator checked Vector2D<int>(Vector2D<T> self);
+    public static explicit operator checked Vector2D<uint>(Vector2D<T> self);
+    public static explicit operator checked Vector2D<long>(Vector2D<T> self);
+    public static explicit operator checked Vector2D<ulong>(Vector2D<T> self);
+    public static explicit operator checked Vector2D<Int128>(Vector2D<T> self);
+    public static explicit operator checked Vector2D<UInt128>(Vector2D<T> self);
+    public static explicit operator checked Vector2D<Half>(Vector2D<T> self);
+    public static explicit operator checked Vector2D<float>(Vector2D<T> self);
+    public static explicit operator checked Vector2D<double>(Vector2D<T> self);
+    public static explicit operator checked Vector2D<decimal>(Vector2D<T> self);
+    public static explicit operator checked Vector2D<Complex>(Vector2D<T> self);
+    public static explicit operator checked Vector2D<BigInteger>(Vector2D<T> self);
+
+    // Cast to System.Numerics.Vector2
+    public static explicit operator Vector2(Vector2D<T> self);
+    public static explicit operator checked Vector2(Vector2D<T> self);
+
+    // Downcast
+
+    // Upcast
+    public static explicit operator Vector3D<T>(Vector2D<T> self);
+    public static explicit operator Vector4D<T>(Vector2D<T> self);
```

Comment:
Upcasting is lossless and *could* be implicit, since the additional elements can be implicitly `0`.

That being said, `ToVector3()` and similar and up being clearer on intent and is how we plan to do it for the BCL types

Reply:
Being explicit was done at @Perksey's request. Note that an implicit cast should match HLSL/GLSL behavior better than a `ToVector3()`-esque method. More discussion may be needed.

Reply:
Looking back at this in GLSL land at least you do have to specify the dimension you're adding iirc, whereas 4D -> 2D should certainly be implicit.

Reply:
@Perksey, why would 4D -> 2D be implicit, since it loses 2 dimensions of data?

Shouldn't in be the inverse, since the new dimensions can be implied `0`?

Reply:
Surely "throw away the bits of data I can't store" is more implied than "assume that I want 0 for the data I don't have". We can make both explicit to be safe.

Reply:
> Surely "throw away the bits of data I can't store" is more implied than "assume that I want 0 for the data I don't have".

Nope

Think about it from the perspective of `Int32->Int64`, this is implicit because the additional data is implied to be zero and it represents, logically, the same value. While `Int32->Int16` is implicit, because it loses data and can't recover it.

In the same way, `Vector2->Vector4` is a widening, and the new data is implied zero (because the identity of a vector is zero, and so new elements take their value from that). You can successfully roundtrip the value back using this.

While `Vector4->Vector2` is a narrowing and may lose data, so you cannot necessarilly recreate the original `Vector4` by doing `Vector4->Vector2->Vector4`.

For a `Matrix`, the implicit widening is that new values on the diagonal are `1`, while others are `0`. This is how `Matrix3x2` ends up working, because its just a space efficient `Matrix3x3` since the last column is `(0, 0, 1)` and never going to be different.

This all ends up working nicely and mathematically when considering the actual algorithm for operations of arbitrary dimensions.

Reply:
good point